### PR TITLE
Issue332 Optimistic Locking Documentation

### PR DESCRIPTION
**Pull request by uweschaefer**

### CODE REVIEW

File: factcast-site/documentation/content/usage/java/optimistic_locking.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+                        // this is only executed once, if publishing succeeded
+                        log.info("Money was transferred.");
+                        
+                    });
+                    
+                });
+
+```
+
+##### Explaination
```
Comment (OOlsen) [inline factcast-site/documentation/content/usage/java/optimistic_locking.md]:
```suggestion
##### Explanation
```

File: factcast-site/documentation/content/usage/java/optimistic_locking.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+First, you tell factcast to record a state according to all events that have either *sourceAccountId* or *targetAccountId* in their list of aggIds and are on namespace *myBankNamespace*. While the namespace is not stricly necessary, it is encouraged to use it - but it depends on your decision on how to use namespaces and group Facts within them.
```
Comment (OOlsen) [inline factcast-site/documentation/content/usage/java/optimistic_locking.md]:
```suggestion
First, you tell factcast to record a state according to all events that have either *sourceAccountId* or *targetAccountId* in their list of aggIds and are on namespace *myBankNamespace*. While the namespace is not strictly necessary, it is encouraged to use it - but it depends on your decision on how to use namespaces and group Facts within them.
```

File: factcast-site/documentation/content/usage/java/optimistic_locking.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+The number of retries is set to *100* here (default is ten, which for many systems is an acceptable default). In essence this means, that the attempt will be executed at max 100 times, before factcast gives up and throws an ***OptimisticRetriesExceededException*** which extends ConcurrentModificationException.
+
+If *interval* is not set, it defaults to 0 with the effect, that the code passed into *attempt* is continously retried without any pause until it either *aborts*, succeeds, or the max number of retries was hit (see above).
```
Comment (OOlsen) [inline factcast-site/documentation/content/usage/java/optimistic_locking.md]:
```suggestion
If *interval* is not set, it defaults to 0 with the effect, that the code passed into *attempt* is continuously retried without any pause until it either *aborts*, succeeds, or the max number of retries was hit (see above).
```

File: factcast-site/documentation/content/usage/java/optimistic_locking.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+                    
+                });
+
+```
+
+##### Explaination
+
+First, you tell factcast to record a state according to all events that have either *sourceAccountId* or *targetAccountId* in their list of aggIds and are on namespace *myBankNamespace*. While the namespace is not stricly necessary, it is encouraged to use it - but it depends on your decision on how to use namespaces and group Facts within them.
+
+The number of retries is set to *100* here (default is ten, which for many systems is an acceptable default). In essence this means, that the attempt will be executed at max 100 times, before factcast gives up and throws an ***OptimisticRetriesExceededException*** which extends ConcurrentModificationException.
```
Comment (OOlsen) [inline factcast-site/documentation/content/usage/java/optimistic_locking.md]: 100 vs ten
100 and 10 (prefer that) or hundred and ten

File: factcast-site/documentation/content/usage/java/optimistic_locking.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+If *interval* is not set, it defaults to 0 with the effect, that the code passed into *attempt* is continously retried without any pause until it either *aborts*, succeeds, or the max number of retries was hit (see above).
+Setting it to *5* means, that before retrying, a 5millisecond wait happens. 
+
+{{< warning >}}<b>WARNING</b>: Setting interval to non-zero makes your code block a thread. The above combination of 100 retries with a 5 msec interval means, that at worst, you code could block <i>longer than half a second</i>.{{< /warning >}}
+
+
+Everything starts with passing a lambda to the *attempt* method. The Lambda is of type
```
Comment (OOlsen) [inline factcast-site/documentation/content/usage/java/optimistic_locking.md]: lambda vs. Lambda

File: factcast-site/documentation/content/usage/java/optimistic_locking.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+```
+
+##### Explaination
+
+First, you tell factcast to record a state according to all events that have either *sourceAccountId* or *targetAccountId* in their list of aggIds and are on namespace *myBankNamespace*. While the namespace is not stricly necessary, it is encouraged to use it - but it depends on your decision on how to use namespaces and group Facts within them.
+
+The number of retries is set to *100* here (default is ten, which for many systems is an acceptable default). In essence this means, that the attempt will be executed at max 100 times, before factcast gives up and throws an ***OptimisticRetriesExceededException*** which extends ConcurrentModificationException.
+
+If *interval* is not set, it defaults to 0 with the effect, that the code passed into *attempt* is continously retried without any pause until it either *aborts*, succeeds, or the max number of retries was hit (see above).
+Setting it to *5* means, that before retrying, a 5millisecond wait happens.
```
Comment (OOlsen) [inline factcast-site/documentation/content/usage/java/optimistic_locking.md]: You use the term "5 msec" in the next warning section.  We should be consistent with the terms.

File: factcast-site/documentation/content/usage/java/optimistic_locking.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+You may probably guess what happens, remembering the above steps. Let's dive into details with a more complex scenario.
+
+##### a complete example
+
+The unavoidable imaginary example, of two BankAccounts and a money transfer between them:
+
+```
+        factcast.lock("myBankNamespace")
+                .on(sourceAccountId,targetAccountId)
+                .optimistic()            // this is optional, defaults to optimistic
```
Comment (OOlsen) [inline factcast-site/documentation/content/usage/java/optimistic_locking.md]: Should we add a hint here? That currently optmistic locking is the only supported mode. For lazy people  that only look at the code examples.  :smile:

File: factcast-site/documentation/content/usage/java/optimistic_locking.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+                        .ns("myBankNamespace")
+                        .type("transfer")
+                        .aggId(sourceAccountId)
+                        .aggId(targetAccountId)
+                        .build("{...}");            
+                    
+                    // register for publishing
+                    return Attempt.publish(toPublish).andThen(()->{
+                        
+                        // this is only executed once, if publishing succeeded
```
Comment (OOlsen) [inline factcast-site/documentation/content/usage/java/optimistic_locking.md]: Can this be mistunderstood? like: If the publishing not suceeded, the function will be excecuted multiple times.

File: factcast-site/documentation/content/usage/java/optimistic_locking.md (hunk 7 above)
Comment (uweschaefer) [reply factcast-site/documentation/content/usage/java/optimistic_locking.md]: done

File: factcast-site/documentation/content/usage/java/optimistic_locking.md (hunk 8 above)
Comment (uweschaefer) [reply factcast-site/documentation/content/usage/java/optimistic_locking.md]: // this is only executed at max once, and only if publishing succeeded

File: factcast-site/documentation/content/usage/java/optimistic_locking.md (hunk 4 above)
Comment (uweschaefer) [reply factcast-site/documentation/content/usage/java/optimistic_locking.md]: done

**Pull request merged**